src/optim/abstraction/hierarchical_abstraction.jl

```diff
@@ -0,0 +1,729 @@
+        from_from = length(path) == 2 ? -2 : path[end - 2]
+        #if the alternating simulation is not yet computed
+        if prev_cell.heuristic_abstraction == nothing
+            build_heuristic_abstraction!(prob, prev_cell)
+        end
+        # build the heuristic if necessary
+        if !haskey(prev_cell.heuristics, from_from)
+            build_heuristic!(prev_cell, from_from)
+        end
+
+        heuristic = prev_cell.heuristics[from_from]
+        _T_L = prev_cell.local_target_set[path[end]]
+        cost = node.parent.lower_bound + get_min_value_heurisitic(heuristic, _T_L)
+        if prev_cell.state != prob.qT
+            cost -= prev_cell.lower_bound
+        else
+            heuristic = prev_cell.heuristics[from_from]
+            _T_L = prev_cell.local_target_set[-1]
+            cost -= get_min_value_heurisitic(heuristic, _T_L)
+        end
+    end
+
+    if curr_cell.state != prob.qT
+        cost += curr_cell.lower_bound
+    else
+        heuristic = curr_cell.heuristics[from]
+        _T_L = curr_cell.local_target_set[-1]
+        cost += get_min_value_heurisitic(heuristic, _T_L)
+    end
+    println("cost:  ", path, " -> ", cost)
+    return node.lower_bound = cost
+end
+
+function set_local_optimizer!(
+    cell,
+    concrete_problem,
+    abstract_system,
+    reference_local_optimizer,
+)
+    optimizers = cell.optimizers
+    concrete_system = concrete_problem.system
+    v = 1.0
+    rec = UT.HyperRectangle(
+        cell.reachable_set.lb - SVector(v, v),
+        cell.reachable_set.ub + SVector(v, v),
+    )
+    d = DO.RectangularObstacles(rec, [])
+    Xdom = DO.GeneralDomainList(
+        reference_local_optimizer.param[:hx];
+        elems = d,
+        periodic = concrete_system.periodic,
+        periods = concrete_system.periods,
+        T0 = concrete_system.T0,
+        fit = true,
+    )
+
+    for (prev_state, _I_L) in cell.local_init_set
+        for (next_state, _T_L) in cell.local_target_set
+            local_concrete_problem = PR.OptimalControlProblem(
+                concrete_problem.system,
+                _I_L[1],
+                _T_L[1],
+                concrete_problem.state_cost,
+                concrete_problem.transition_cost,
+                concrete_problem.time,
+            )
+            local_optimizer = copy(reference_local_optimizer)
+            LazyAbstraction.set_optimizer!(
+                local_optimizer,
+                local_concrete_problem,
+                DO.get_grid(abstract_system.symmodel.Udom);
+                Xdom = Xdom,
+            )
+            local_optimizer.abstract_system_heuristic = cell.heuristic_abstraction
+            optimizers[(prev_state, next_state)] = local_optimizer
+        end
+    end
+end
+
+function BB.compute_upper_bound!(prob::HierarchicalProblem, node::BB.Node)
+    global fig
+    path = node.elem
+    cells = prob.cells
+    if path[end] != prob.qT
+        return
+    end
+    n_threads = 4
```

Comment:
yes, I think it's when you launch julia that you should choose the number of threads allowed. By default, it will use a single thread to execute this code, but these lines are allowed to be multithreaded if julia has been launched in such a way that: 'julia --threads 4'. At least, that's what I've done locally.

Reply:
OK I didn't know that, is there a way to change this in the REPL itself?

Reply:
I don't know, but it would be cool, but looking around (I didn't search very long :) ) I couldn't find a way to force it to use a specific number of threads once I launch: "julia".

Reply:
Me neither haha